swarm/network/stream/streamer_test.go

```diff
@@ -921,3 +926,159 @@ func TestMaxPeerServersWithoutUnsubscribe(t *testing.T) {
+	}
+	binMap[1] = []string{
+		"28f0bc1b44658548d6e05dd16d4c2fe77f1da5d48b6774bc4263b045725d0c19",
+		"2690a910c33ee37b91eb6c4e0731d1d345e2dc3b46d308503a6e85bbc242c69e",
+	}
+	binMap[2] = []string{
+		"4a45f1fc63e1a9cb9dfa44c98da2f3d20c2923e5d75ff60b2db9d1bdb0c54d51",
+		"4d72a04ddeb851a68cd197ef9a92a3e2ff01fbbff638e64929dd1a9c2e150112",
+	}
+	binMap[3] = []string{
+		"646e9540c84f6a2f9cf6585d45a4c219573b4fd1b64a3c9a1386fc5cf98c0d4d",
+	}
+	binMap[4] = []string{
+		"7656caccdc79cd8d7ce66d415cc96a718e8271c62fb35746bfc2b49faf3eebf3",
+		"76d1e83c71ca246d042e37ff1db181f2776265fbcfdc890ce230bfa617c9c2f0",
+		"769ce86aa90b518b7ed382f9fdacfbed93574e18dc98fe6c342e4f9f409c2d5a",
+	}
+	binMap[5] = []string{
+		"7a48f75f8ca60487ae42d6f92b785581b40b91f2da551ae73d5eae46640e02e8",
+	}
+	binMap[6] = []string{
+		"7cbd42350bde8e18ae5b955b5450f8e2cef3419f92fbf5598160c60fd78619f0",
+	}
+
+	// create the pivot's kademlia
+	addr := common.FromHex(pivotAddr)
+	k := network.NewKademlia(addr, network.NewKadParams())
+
+	// construct the peers and the kademlia
+	for _, binaddrs := range binMap {
+		for _, a := range binaddrs {
+			addr := common.FromHex(a)
+			k.On(network.NewPeer(&network.BzzPeer{BzzAddr: &network.BzzAddr{OAddr: addr}}, k))
+		}
+	}
+
+	// TODO: check kad table is same
+	// currently k.String() prints date so it will never be the same :)
+	// --> implement JSON representation of kad table
+	log.Debug(k.String())
+
+	// simulate that we would do subscriptions: just store the bin numbers
+	fakeSubscriptions := make(map[string][]int)
+	// define the function which should run for each connection
+	// instead of doing real subscriptions, we just store the bin numbers
+	requestSubscriptionFunc := func(p *network.Peer, bin uint8, subs map[enode.ID]map[Stream]struct{}) bool {
+		// get the peer ID
+		peerstr := fmt.Sprintf("%x", p.Over())
+		// create the array of bins per peer
+		if _, ok := fakeSubscriptions[peerstr]; !ok {
+			fakeSubscriptions[peerstr] = make([]int, 0)
+		}
+		// store the (fake) bin subscription
+		log.Debug(fmt.Sprintf("Adding fake subscription for peer %s with bin %d", peerstr, bin))
+		fakeSubscriptions[peerstr] = append(fakeSubscriptions[peerstr], int(bin))
+		return true
+	}
+	// create just a simple Registry object in order to be able to call...
+	r := &Registry{}
+	// ...the requestPeerSubscriptions function, which contains the logic for subscriptions
+	r.requestPeerSubscriptions(k, nil, requestSubscriptionFunc)
+	// calculate the kademlia depth
+	kdepth := k.NeighbourhoodDepth()
+
+	// now, check that all peers have the expected (fake) subscriptions
+	// iterate the bin map
+	for bin, peers := range binMap {
+		// for every peer...
+		for _, peer := range peers {
+			// ...get its (fake) subscriptions
+			fakeSubs := fakeSubscriptions[peer]
+			// if the peer's bin is below the kademlia depth...
+			if bin < kdepth {
+				// (iterate all (fake) subscriptions)
+				for _, subbin := range fakeSubs {
```

Comment:
If there's just one maybe don't need iterator

Reply:
Hmmm sure but why introduce more complexity by having to check if it's just one, and if there are more than one, still have an iterator? I vote to keep it as is

Reply:
Ok fine.